### ISSUE TITLE
WebGPU Performance Issues

### ISSUE THREAD

**Issue by vladmandic** [tensorflow/tfjs#5689]
i just tried new `tfjs-backend-webgpu` *0.0.1-alpha.8* on `tfjs` *3.9.0*   
*environment: chrome 96 canary on windows 11*

first, great job on adding tons of new ops - from perspective of supported kernel ops, `webgpu` is becoming usable!

however, switch to WGSL is anything but useful so far - it comes as a **major performance degredation**  

overall, `webgpu` has  gotten slower than `webgl`  
(and `webgl` itself has become significantly slower since `tfjs` 3.4.0 - this is discussed separately in several open issues)  

not to mention that new work that has gone into `webgl` to make it manageable (enable uniforms) has no effect on `webgpu`

comparing warmup times  
(fyi, my app by default uses 8 simple models running in parallel - total models size is actually tiny, below 30mb):

- `webgl` *(default settings)*
  > **14 sec** (double the value with uniforms enabled)
- `webgl` with `WEBGL_PACK_DEPTHWISECONV=false` and `WEBGL_USE_SHAPES_UNIFORMS=true`
  > **7 sec** (pretty good)
- `webgpu` *(default settings)*
  > **25 sec** (this is incredibily slow)
- `webgpu` with `WEBGPU_USE_GLSL=true`
  > **15 sec** (already slower than webgl)
- `wasm` (no real warmup, included for refrerence only)
  > **2 sec**

*imo, when developing new backend, goal should be that its better than the previous one - not just that it passes unit tests
if `webgpu` is not significantly improved, it will be a d.o.a. once released*

cc @qjia7 and @xhcao due to work on webgpu  
cc @pyu10055 as assignee on webgl performance degradation issue

**Comment by gyagp** [comment]
@vladmandic Thanks for the good comments and data, as always! 
Chrome 94 was released on Sep 21, with WebGPU Origin Trial support. This means in addition to Chrome Canary, we may use Chrome Stable (still need option --enable-unsafe-webgpu) for WebGPU experiment now. Unfortunately,  Chrome decided not to support GLSL anymore for WebGPU  (changes happened in master so all the release channels would be impacted, including Canary and Stable), so WGSL is the only one that can be consumed now. We always align well with WebGPU development (My team also heavily contributes to WebGPU spec, CTS and Chrome impl) and started the TFJS GLSL to WGSL transition in June. After fixing many critical perf issues in Chrome (e.g., workgroup memory init perf regression) together with Google and working around perf issues in TFJS (e.g., hardware limits), we finished the transition after 3+ months of work.
Internally we have daily track of performance against almost all the workloads defined in TFJS e2e benchmarks. Before switching to WGSL, we double-checked there was no performance regression regarding to warmup time and run time. For sure, due to resources, we could only cover very limited platforms (Actually only Intel Coffee Lake and Tiger Lake are under daily test), and very limited workloads. We'd like to hear more details from your side (e.g., hardware configuration) to understand the regression. We'll investigate right after our holidays (We are off from Oct 1 to 7 for National Day Holidays). 
BTW,
1. The uniform idea was already implemented in WebGPU backend. Google thought it great, so we're bringing it to WebGL backend.
2. Comparing with WebGL, compiled shaders couldn't be cached in Chrome. We already raised this implementation issue to Chrome and it's going to take a while for its implementation (not easy). 

Thanks again for your valuable feedback, hopes to hear more details from your side about warmup regression (e.g., hardware configuration), and looks forward to more collaborations in the future!

**Comment by qjia7** [comment]
@vladmandic Can you put the Inception v4 model somewhere that I can access? It seems that `http://wyse:10010/models/imagenet/inception-v4/model.json` is in your local server. The webgpu warning seems like a bug in our implementation.

And for your demo app, I can reproduce the bad performance for webgpu. Thanks for the reporting. I will take a look.

**Comment by vladmandic** [comment]
@qjia7 did you have a chance to look at the `webgpu` latency issues i've mentioned? tests above is from 22 days ago.

**Comment by qjia7** [comment]
@vladmandic Sorry, I can't reproduce the latency issue you mentioned. From the code snippet you paste, it basically does nothing and is executed instantaneously in my side.
```
  const numIterations = 50;
  const arr = new Uint8Array(imageData?.data.buffer); // input data in my case is 4k imageData, but can be any dataset
  const t0 = performance.now();
  for (let i = 0; i < numIterations; i++) {
    const rgba = tf.tensor(arr, [imageData.width, imageData.height, 4], 'int32'); // create rgba tensor
    const rgb = tf.slice3d(rgba, [0, 0, 0], [-1, -1, 3]); // strip alpha channel
    const tensor = tf.expandDims(rgb, 0); // create standard image tensor [1, height, width, 3]
    // const data = await tensor.array(); // download data from gpu
    tf.dispose([rgba, rgb, tensor]); // just dispose everything
  }
  const t1 = performance.now();
  const avgTime = Math.round((t1 - t0) / numIterations);
  console.log({ backend: tf.getBackend(), average: avgTime });
```

Some update for the warmup time:

- In tfjs level, we are reducing some shader variants firstly from binary ops #5791. Will start others when we get each operator's exact shader compilation time. Some challenges for us are that 1) it's hard to measure each operator's shader compilation time. Details can be found [here](https://bugs.chromium.org/p/chromium/issues/detail?id=1265616), [here](https://github.com/gpuweb/gpuweb/issues/2236). 2) browser devtools also [lack this support](https://bugs.chromium.org/p/dawn/issues/detail?id=1167). Fortunately, chrome just supported it in chrome://tracing. We can continue this work once it's ready in canary.

- In webgpu level, google developer has started the work to [cache pipeline in memory](https://bugs.chromium.org/p/dawn/issues/detail?id=549).

**Comment by vladmandic** [comment]
@qjia7 thanks for the update!  

and the description of the chromium queue handling is sounds like it could be the same root cause for what i'm seeing as extreme latency issues  

for reproduction, im guessing your test failed since `imageData` was empty, but you can even use `tf.zeros` to reproduce  
no model needed, nothing - just a trivial loop  

here's a live link: <https://vladmandic.github.io/tfjs-utils/src/latency-issue.html>  

and output on my notebook:

```text
user agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/97.0.4689.0 Safari/537.36
tf version: 3.11.0-20211102
backend: webgl | total time: 4556 ms | average time: 91 ms
backend: webgpu | total time: 13058 ms | average time: 261 ms
```

basically, for any "real" work, `webgpu` is really fast - but for simple stuff that is done in js code outside of the model, latency is a killer - its 300% slower than `webgl`  
unfortunately, in real-world, any model inference is followed by some post-processing in js and that is where this impact becomes as showstopper  

setting `tf.ENV.set('WEBGPU_DEFERRED_SUBMIT_BATCH_SIZE', 0)` reduces the latency by 50%,  
but its still nowhere as fast as `webgl`

**Comment by qjia7** [comment]
@vladmandic Thanks for your live case. I can reproduce it now. After debugging, I find the time mainly costs on `queue.writeBuffer`. It seems that this API needs to be optimized in browser for big data uploading. I reported a bug to chromium https://bugs.chromium.org/p/chromium/issues/detail?id=1266727.

**Comment by qjia7** [comment]
@vladmandic Jiawei in our team has fixed the `queue.writeBuffer` issue in chromium. You can retest your example https://vladmandic.github.io/tfjs-utils/src/latency-issue.html with latest chrome canary (--enable-unsafe-webgpu). The webgpu backend becomes much faster than before. In my machine, the latency is not that obvious now. We are also trying to use `mapAsync` instead of `writeBuffer`, which shows better perf than webgl with your example. But it brings another [issue](https://github.com/tensorflow/tfjs/pull/5928#discussion_r765587353). It's still in discussion, but hope we can provide the most performant solution soon.

And for the long warmup time, we drafted the prototype for the parallel compilation, showing almost 4x speedup for the warmup time. Currently, we are discussing how to expose this capability uniformly between webgl and webgpu.  Will keep you updated.

**Comment by vladmandic** [comment]
Thanks @qjia7!

I've tested with Chrome 99 and latency is gone - **WebGPU** now performs on-par with **WebGL**

I'm looking forward to other proposed changes (once issue is resolved) for `mapAsync` in #5928
And I'm guessing you're still discussing how to align changes from #5826 with changes from #5815 ?

Anyhow, I'm closing this issue as resolved...
